### PR TITLE
[train] add TrainControllerState metrics

### PR DESCRIPTION
**Pull request by matthewdeng**
## Summary
1. Refactors out the metrics handling logic into an internal metrics module.
2. Implements a new metric for tracking TrainController state over time, similar to Ray Core's Task/Actor state metrics.
3. Add a Grafana dashboard panel that shows the states.

## Refactoring

All the metrics handling logic is now abstracted away in an internal `metrics` module.
```
python/ray/train/v2/_internal/metrics
├── __init__.py
├── base.py
├── controller.py
└── worker.py
```

As a result, the `ControllerMetricsCallback` and `WorkerMetricsCallback` can now be thin layers that map the callback events to calls to `MetricsTracker.update`.

```python
class Metric:
    def start(self):
        ...
    
    def record(self, tags: Dict[str, str], value: T):
        ...
    
    def get_value(self, tags: Dict[str, str]) -> T:
        ...
    
    def reset(self):
        ...
    
    def shutdown(self):
        ...
```

## TrainControllerState

Added a new metric to track `TrainControllerState`. 

Its key is defined ("ray_train_run_name", "ray_train_controller_state"), and this should be taken into consideration when defining the visualization for this metric.

## Example

Ran this locally, with a small modification to print out calls to `ray.util.metrics.Gauge.set()`.

**Repro script**
```python
import time
from ray.train.torch import TorchTrainer

def train_func():
    time.sleep(10)

trainer = TorchTrainer(train_func)
trainer.fit()
```

**Logs** 
```
(TrainController pid=39720) SET train_controller_state 1 {'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36', 'ray_train_controller_state': 'INITIALIZING'}
(TrainController pid=39720) SET train_controller_state 0 {'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36', 'ray_train_controller_state': 'INITIALIZING'}
(TrainController pid=39720) SET train_controller_state 1 {'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36', 'ray_train_controller_state': 'SCHEDULING'}
(TrainController pid=39720) Started training worker group of size 1: 
(TrainController pid=39720) - (ip=127.0.0.1, pid=39728) world_rank=0, local_rank=0, node_rank=0
(RayTrainWorker pid=39728) Setting up process group for: env:// [rank=0, world_size=1]
(TrainController pid=39720) SET train_worker_group_start_total_time_s 1.9262119578197598 {'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36'}
(TrainController pid=39720) SET train_controller_state 0 {'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36', 'ray_train_controller_state': 'SCHEDULING'}
(TrainController pid=39720) SET train_controller_state 1 {'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36', 'ray_train_controller_state': 'RUNNING'}
(TrainController pid=39720) SET train_controller_state 0 {'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36', 'ray_train_controller_state': 'RUNNING'}
(TrainController pid=39720) SET train_controller_state 1 {'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36', 'ray_train_controller_state': 'FINISHED'}
(TrainController pid=39720) SET train_worker_group_shutdown_total_time_s 0.004093542229384184 {'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36'}
(TrainController pid=39720) SET train_worker_group_start_total_time_s 0.0 {'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36'}
(TrainController pid=39720) SET train_worker_group_shutdown_total_time_s 0.0 {'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36'}
(TrainController pid=39720) SET train_controller_state 0 {'ray_train_controller_state': 'INITIALIZING', 'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36'}
(TrainController pid=39720) SET train_controller_state 0 {'ray_train_controller_state': 'SCHEDULING', 'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36'}
(TrainController pid=39720) SET train_controller_state 0 {'ray_train_controller_state': 'RUNNING', 'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36'}
(TrainController pid=39720) SET train_controller_state 0 {'ray_train_controller_state': 'FINISHED', 'ray_train_run_name': 'ray_train_run-2025-05-07_17-56-36'}
```

## Dashboard

Added a new dashboard panel that shows the state.

Note that the dashboard has 15 second increments, so states that are shorter than this period may not show up.
image

Repro script:
```python
import time

from ray.train import RunConfig
from ray.train.backend import Backend, BackendConfig
from ray.train.v2.api.data_parallel_trainer import DataParallelTrainer


class SlowBackend(Backend):
    def on_start(self, worker_group, backend_config):
        time.sleep(100)

    def on_shutdown(self, worker_group, backend_config):
        time.sleep(100)


class SlowBackendConfig(BackendConfig):
    @property
    def backend_cls(self):
        return SlowBackend


def train_func():
    time.sleep(100)


run_config = RunConfig(name="slow-backend-run")


trainer = DataParallelTrainer(
    train_func, backend_config=SlowBackendConfig(), run_config=run_config
)
trainer.fit()
```

### CODE REVIEW

**Review by justinvyu — COMMENTED** [review]
The "additional tags" argument seems like a confusing implementation detail of the state update metric. Wondering if a pattern like this is simpler:

```python
class EnumMetric(Metric):
    def __init__(self, enum_cls):
        self._lock = nullcontext()

    def create_gauge(self):
        self._gauge = Gauge(...)
    
    def set_lock(self, lock):
        self._lock = lock

    def update(self, new_state: Enum):
        with self.lock:
            self._value = new_state

    def push(self):
        with self._lock:
            for option in Enum.options:
                if self._value == option:
                    self._gauge.set(1, {"ray_train_controller_state": self._value})
                else:
                    self._gauge.set(0, {"ray_train_controller_state": self._value})


controller_state_metric = EnumMetric(ControllerState, ...)
tracker = MetricsTracker([controller_state_metric])  # This creates and sets the lock on all metrics passed in
controller_state_metric.update(curr_controller_state)
```

File: python/ray/train/v2/_internal/metrics/base.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        self._values = {}
+
+    def update(self, metric: Metric, additional_tags: Dict[str, str], value: Any):
+        """Update a metric value with associated tags.
+
+        Args:
+            metric: A Metric instance
+            additional_tags: Dictionary of tag key-value pairs to be merged with the base tags
+            value: The value to update the metric with. The value will be added to the existing value
+                for the metric-tags combination, or set if the metric-tags combination does not exist.
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/train/v2/_internal/metrics/base.py:54]: Ooh this is a great callout. I was overindexing a bit on this particular new metric that I added.

File: python/ray/train/v2/_internal/metrics/controller.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    default=0.0,
+    description="Cumulative time in seconds to shutdown worker groups in the Train job.",
+)
+
+TRAIN_CONTROLLER_STATE = ControllerMetric(
+    name="train_controller_state",
+    type=int,
+    default=0,
+    description="The current state of the controller",
+    tag_keys=CONTROLLER_TAG_KEYS + (CONTROLLER_STATE_TAG_KEY,),
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/train/v2/_internal/metrics/controller.py:34]: This is meant to validate the tag keys that need to be specified when logging this metric. There's actually validation that happens at the lower level if the tags aren't passed in, but I can add a quick validation at the `update`/`record` layer as well so it fast-fails!

File: python/ray/train/v2/_internal/metrics/base.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                f"Expected: {self.tag_keys}, got: {list(tags.keys())}"
+            )
+
+
+class MetricsTracker:
+    """Tracks metric values for a set of defined metrics."""
+
+    def __init__(self, metrics: Dict[str, Metric], base_tags: Dict[str, str]):
+        self._metrics = metrics
+        self._base_tags = base_tags
```
Comment (justinvyu) [inline python/ray/train/v2/_internal/metrics/base.py:111]: I don't think we need MetricsTracker anymore if we don't need to coordinate the lock across metrics. Right now it just serves as a "metrics list" which also requires that all metrics have the same report interface/implementation.

Here's what I was thinking:

```python
class Metric:
    def record(self, value: T):
        pass

class TimeMetric(Metric):
    def record(self, value: T):
        self._value += value
        gauge.set(self._value, self.tags)

class EnumMetric(Metric):
    def record(self, value: T):
        self._value = value
        for option in list(self.type):
            gauge.set(..., self.tags | {"...": str(option)})


WORKER_METRICS = {
    "train_report_total_blocked_time_s": TimeMetric(...)
}

class WorkerMetricsCallback(WorkerCallback):
    def after_train_context_init(self):
        # No need to abstract away the for loop on worker metrics.
        for metric in WORKER_METRICS.values():
            metric.start()

    def on_report():
        # Just use the metric objects directly rather than go through the metrics tracker
        WORKER_METRICS["train_report_total_blocked_time_s"].record(time_elapsed_s)
```

**Review by justinvyu — COMMENTED** [review]
Is gauge.set() a very cheap operation? Was there a point to doing the background thread loop in the first place?

File: python/ray/train/v2/_internal/metrics/base.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        """
+        self._current_value += value
+        self._gauge.set(self._current_value, self._base_tags)
+
+    def get_value(self) -> float:
+        return self._current_value
+
+    def reset(self):
+        self._gauge.set(self._default, self._base_tags)
+        self._current_value = 0.0
```
Comment (justinvyu) [inline python/ray/train/v2/_internal/metrics/base.py:97]: should we set it to default?

Comment (justinvyu) [reply python/ray/train/v2/_internal/metrics/base.py:97]: or maybe just remove "Default" from the base class since it's not used there

File: python/ray/train/v2/_internal/metrics/base.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        Args:
+            enum_value: The enum value to get the value for
+
+        Returns:
+            The value for the enum value
+        """
+        if enum_value == self._current_value:
+            return 1
+        else:
+            return 0
```
Comment (justinvyu) [inline python/ray/train/v2/_internal/metrics/base.py:151]: 😎 

```suggestion
        return int(enum_value == self._current_value)
```

File: python/ray/train/v2/tests/test_metrics.py [hunk 6]
```diff
@@ -9,43 +9,120 @@
...
     def __init__(self, name: str, description: str, tag_keys: tuple = ()):
-        self._value: float = 0.0
+        self._values: dict[set[str], float] = {}
 
     def set(self, value: float, tags: dict):
-        self._value = value
+        self._values[frozenset(tags.items())] = value
+
+    def get(self, tags: dict):
+        return self._values.get(frozenset(tags.items()))
```
Comment (justinvyu) [inline python/ray/train/v2/tests/test_metrics.py:31]: This is not part of the Gauge API right? And I don't think we use this, let's remove.

File: python/ray/train/v2/tests/test_metrics.py [hunk 7]
```diff
@@ -9,43 +9,120 @@
...
 
-    def get(self):
-        return self._value
+
+class MockTrainControllerState:
+    def __init__(self, state_type):
+        self._state_type = state_type
```
Comment (justinvyu) [inline python/ray/train/v2/tests/test_metrics.py:36]: just use TrainControllerState. we're mocking out the private attribute which is a bit weird here

File: python/ray/train/v2/tests/test_metrics.py [hunk 8]
```diff
@@ -9,43 +9,120 @@
...
+        time_index = time_index + 1
+        return value
+
+    monkeypatch.setattr(
+        ray.train.v2._internal.callbacks.metrics, "time_monotonic", mock_time
+    )
+
+
+def test_time_metric(monkeypatch):
+    """Test TimeMetric functionality."""
```
Comment (justinvyu) [inline python/ray/train/v2/tests/test_metrics.py:59]:
```suggestion
```

File: python/ray/train/v2/tests/test_metrics.py [hunk 9]
```diff
@@ -9,43 +9,120 @@
...
+    monkeypatch.setattr(ray.train.v2._internal.metrics.base, "Gauge", MockGauge)
```
Comment (justinvyu) [inline python/ray/train/v2/tests/test_metrics.py:60]: make this a fixture

File: python/ray/train/v2/tests/test_metrics.py [hunk 10]
```diff
@@ -9,43 +9,120 @@
...
 
 
-def mock_on_worker_group_event(self, value: float, event: str):
-    """Mock function to set the value of the worker group event"""
-    if event == "start":
-        self._metrics.train_worker_group_start_total_time_s += value
-    elif event == "shutdown":
-        self._metrics.train_worker_group_shutdown_total_time_s += value
+def test_enum_metric(monkeypatch):
+    """Test EnumMetric functionality."""
```
Comment (justinvyu) [inline python/ray/train/v2/tests/test_metrics.py:83]:
```suggestion
```

File: python/ray/train/v2/tests/test_metrics.py [hunk 11]
```diff
@@ -9,43 +9,120 @@
...
+    assert metric.get_value(TestEnum.A) == 0
+    assert metric.get_value(TestEnum.B) == 0
+    assert metric.get_value(TestEnum.C) == 0
 
 
 def test_worker_metrics_callback(monkeypatch):
-    monkeypatch.setattr(WorkerMetricsCallback, "LOCAL_METRICS_PUSH_INTERVAL_S", 0.05)
-    monkeypatch.setattr(WorkerMetricsCallback, "on_report", mock_on_report)
+    monkeypatch.setattr(ray.train.v2._internal.metrics.base, "Gauge", MockGauge)
+    mock_time_monotonic(monkeypatch, [0.0, 1.0, 10.0, 12.0])
```
Comment (justinvyu) [inline python/ray/train/v2/tests/test_metrics.py:124]: so is this a total of 3 seconds? (0-1, 10-12). Can you add a comment about this, or make it more clear by having the list be tuples of start,end?

File: python/ray/train/v2/tests/test_metrics.py [hunk 12]
```diff
@@ -54,42 +131,81 @@ def test_worker_metrics_callback(monkeypatch):
...
+    with callback.on_report():
+        pass
+    assert (
+        callback._metrics[WorkerMetrics.REPORT_TOTAL_BLOCKED_TIME_S].get_value() == 3.0
+    )
 
 
 def test_controller_metrics_callback(monkeypatch):
+    monkeypatch.setattr(ray.train.v2._internal.metrics.base, "Gauge", MockGauge)
+    mock_time_monotonic(monkeypatch, [0.0, 1.0, 10.0, 12.0])
```
Comment (justinvyu) [inline python/ray/train/v2/tests/test_metrics.py:157]: also test shutdown logic with before_controller_shutdown and before_shutdown for worker and controller metrics callback

File: python/ray/train/v2/_internal/metrics/controller.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+            name=name,
+            description=description,
+            base_tags=base_tags,
+        )
+
+    @classmethod
+    def _create_controller_state_metric(
+        cls, base_tags: Dict[str, str]
+    ) -> EnumMetric[TrainControllerStateType]:
+        return EnumMetric[TrainControllerStateType](
```
Comment (justinvyu) [inline python/ray/train/v2/_internal/metrics/controller.py:37]: cool

File: python/ray/train/v2/_internal/metrics/base.py (hunk 4 above)
Comment (matthewdeng) [reply python/ray/train/v2/_internal/metrics/base.py:97]: yes good call, I will remove since `Metric` is now abstract!

Comment (matthewdeng) [reply python/ray/train/v2/_internal/metrics/base.py:97]: Oh actually I will keep it and convert to use default because I think it's nice to guarantee this in the logic of `get_value`

**Review by justinvyu — APPROVED** [review]
🆒
